overlord/ifacestate/apparmorprompting/common/patterns.go

```diff
@@ -0,0 +1,478 @@
+			currSubpatternStart = indexOfNextRune(reader)
+			newPrefixes := make([]string, 0, len(currPrefixes)*len(groupExpanded))
+			for _, prefix := range currPrefixes {
+				for _, suffix := range groupExpanded {
+					newPrefixes = append(newPrefixes, prefix+infix+suffix)
+				}
+			}
+			currPrefixes = newPrefixes
+			continue
+		}
+		if r == ',' || r == '}' {
+			suffix := prevStringFromIndex(reader, currSubpatternStart)
+			newExpanded := make([]string, len(expanded), len(expanded)+len(currPrefixes))
+			copy(newExpanded, expanded)
+			expanded = newExpanded
+			for _, prefix := range currPrefixes {
+				expanded = append(expanded, prefix+suffix)
+			}
+			currPrefixes = defaultPrefixes
+			currSubpatternStart = indexOfNextRune(reader)
+		}
+		if r == '}' {
+			return expanded, nil
+		}
+	}
+	// Group missing closing '}' character, so return an error.
+	origPatternBuf := make([]byte, reader.Size())
+	reader.ReadAt(origPatternBuf, 0)
+	return nil, fmt.Errorf(`invalid path pattern: unmatched '{' character: %q`, string(origPatternBuf))
+}
+
+var (
+	duplicateSlashes    = regexp.MustCompile(`(^|[^\\])/+`)
+	charsDoublestar     = regexp.MustCompile(`([^/\\])\*\*+`)
+	doublestarChars     = regexp.MustCompile(`([^\\])\*\*+([^/])`)
+	duplicateDoublestar = regexp.MustCompile(`/\*\*(/\*\*)+`) // relies on charsDoublestar running first
+	starsAnyMaybeStars  = regexp.MustCompile(`([^\\])\*+(\?\**)+`)
+)
+
+func cleanPattern(pattern string) string {
+	pattern = duplicateSlashes.ReplaceAllString(pattern, `${1}/`)
+	pattern = charsDoublestar.ReplaceAllString(pattern, `${1}*`)
+	pattern = doublestarChars.ReplaceAllString(pattern, `${1}*${2}`)
+	pattern = duplicateDoublestar.ReplaceAllString(pattern, `/**`)
+	pattern = starsAnyMaybeStars.ReplaceAllStringFunc(pattern, func(s string) string {
+		deleteStars := func(r rune) rune {
+			if r == '*' {
+				return -1
+			}
+			return r
+		}
+		return strings.Map(deleteStars, s) + "*"
+	})
+	if strings.HasSuffix(pattern, "/**/*") {
+		// Strip trailing "/*" from suffix
+		return pattern[:len(pattern)-len("/*")]
+	}
+	return pattern
+}
+
+type priorityType int
+
+const (
+	worstPriority priorityType = iota
+	priorityGlobDoublestar
+	priorityTerminalDoublestar
+	priorityDoublestar
+	priorityGlob
+	prioritySinglestar
+	prioritySingleChar
+	priorityTerminated
+	priorityLiteral
+)
+
+type nextPatternsContainer struct {
+	currPriority     priorityType
+	nextPatternsList []*strings.Reader
+}
+
+func (np *nextPatternsContainer) addWithPriority(priority priorityType, reader *strings.Reader) {
+	if priority < np.currPriority {
+		return
+	}
+	if priority > np.currPriority {
+		np.nextPatternsList = np.nextPatternsList[:0]
+		np.currPriority = priority
+	}
+	np.nextPatternsList = append(np.nextPatternsList, reader)
+}
+
+func (np *nextPatternsContainer) nextPatterns() []*strings.Reader {
+	return np.nextPatternsList
+}
+
+// GetHighestPrecedencePattern determines which of the given path patterns is
```

Comment:
out of curiosity, do you recall what apparmor_parser does if there are potentially overlapping file rules?

Reply:
AppArmor denies if there is any deny rule which matches the given path. But we don't want that behavior here -- we want the most specific rule to have precedence here. For example, a user could grant read access to `~/Documents` but deny read access to `~/Documents/private`.